src/uct/ib/base/ib_iface.c

```diff
@@ -1173,6 +1174,91 @@ static void uct_ib_iface_set_path_mtu(uct_ib_iface_t *iface,
     }
 }
 
+uint8_t uct_ib_iface_config_select_sl(const uct_ib_iface_config_t *ib_config)
+{
+    ucs_assert((ib_config->sl <= UCT_IB_SL_MAX) ||
+               (ib_config->sl == UCS_ULUNITS_AUTO));
+    return (ib_config->sl == UCS_ULUNITS_AUTO) ? 0 : (uint8_t)ib_config->sl;
+}
+
+ucs_status_t
+uct_ib_iface_select_sl(const uct_ib_iface_config_t *ib_config,
+                       ucs_ternary_value_t ar_enable, uint16_t ooo_sl_mask,
+                       int ooo_sl_mask_is_detected,
+                       const char *dev_port_name, uint8_t *sl_p)
+{
+    const char *sl_ar_support_str = (ar_enable == UCS_YES) ? "with" : "without";
+    ucs_status_t status           = UCS_OK;
+    ucs_string_buffer_t sl_mask_str;
+    char sl_str[8];
+    char ar_enable_str[8];
+    uint8_t sl_from_mask;
+
+    ucs_assert(ooo_sl_mask_is_detected || (ooo_sl_mask == 0));
+
+    if ((ar_enable == UCS_NO) && ooo_sl_mask_is_detected) {
+        ooo_sl_mask = ~ooo_sl_mask;
```

Comment:
does it mean it will fail if
  AR_ENABLE=NO, ooo_sl_mask_is_detected is false, ooo_sl_mask = 0, sl != auto?

Reply:
yep, we have a test for this